libs/base/Control/Relation.idr

```diff
@@ -0,0 +1,12 @@
+||| Properties of binary relations
+module Control.Relation
+
+%default total
+
+public export
+interface Reflexive (ty : Type) (rel : ty -> ty -> Type) where
+  reflexive : {x : ty} -> rel x x
+
+public export
+interface Transitive (ty : Type) (rel : ty -> ty -> Type) where
+  transitive : {x, y, z : ty} -> rel x y -> rel y z -> rel x z
```

Comment:
From what I see, you have to supply these implicit arguments explicitly anyway. I believe practicality should be a guideline as to which arguments should be implicit and which explicit. If the compiler can't infer these by itself, then in my opinion they should be explicit. These curly brackets are noisy and I'd prefer to avoid writing them. If you can't do that most of the time, probably the argument shouldn't be implicit after all.

Reply:
These "explicit implicit" arguments are required because implementations may need to case-split on them. For example,

```idris
Reflexive Nat LTE where
  reflexive {x = Z} = LTEZero
  reflexive {x = S k} = LTESucc $ reflexive {x = k}
```

That's not a `Transitive` instance, but you get the idea.

Reply:
I'm not complaining about their existence. They're obviously required. What bothers me is the necessity of writing these curly braces when calling a function. Pattern matching on them is not that bad, because you do it only once and besides pattern matching when implementing a function is often unavoidable. My point is about calling the function. If the compiler can't infer the argument most of the time, then it should probably be explicit. You made a good point below, though.

Reply:
On the other hand in my experience, having all arguments implicit damages the readability. Suppose you indeed don't need to specify any of them, because the compiler can work all of them out on its own. However, when human reads that code, trying to understand it, he or she will often have a hard time figuring out, reflexivity of which relation and over which value is important here. So specifying the value is usually a nice hint for a human reader, event if it's not exactly relevant to the compiler. Moreover, even if the compiler can figure it out having the complete code is one thing, but these arguments also help it make sense of things in the presence of holes.

Reply:
I consider the braced arguments as temporary workarounds. At some point in the future it may be possible to go through and delete all of them in one fell swoop.